.github/workflows/main.yml

```diff
@@ -7,85 +8,90 @@ on:
 
 jobs:
   validation:
-    name: "Validation"
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
       - uses: gradle/wrapper-validation-action@v1
 
   checks:
-    runs-on: macos-latest
+    runs-on: ubuntu-latest
     steps:
-
-      - name: Project Checkout
-        uses: actions/checkout@v2
-
-      - name: Setup JDK 1.8
-        uses: actions/setup-java@v2
+      - uses: actions/checkout@v3
+      - uses: actions/setup-java@v3
         with:
           java-version: 8
           distribution: 'zulu'
-
+      - uses: gradle/gradle-build-action@v2
       - name: Build project
         run: ./gradlew build --stacktrace
+      - name: Assemble Test APK
+        run: ./gradlew assembleAndroidTest --stacktrace
 
   instrumentation-tests:
-    name: Instrumentation tests
     runs-on: macos-latest
     timeout-minutes: 30
     strategy:
-      # Allow tests to continue on other devices if they fail on one device.
-      fail-fast: false
       matrix:
-        arch: [x86_64]
-        target: [google_apis]
-        channel: [stable]
-        api-level:
-          - 21
-          - 23
-          - 26
-         # - 29 flaky
-        include:
-          - arch: x86
+        api-level: [ 16, 19, 21, 23, 26, 29 ]
+        target: [ default, google_apis ]
+        arch: [ x86 ]
+        exclude:
+          - target: google_apis
             api-level: 16
-            target: google_apis
-            channel: stable
-          - arch: x86
+          - target: google_apis
             api-level: 19
-            target: google_apis
-            channel: stable
-        # Failing (something about permissions maybe?
-        #  - arch: x86
-        #    api-level: 30
-        #    target: aosp_atd
-        #    channel: canary
-
+          - target: google_apis
+            api-level: 21
+          - target: google_apis
+            api-level: 23
+          - target: google_apis
+            api-level: 26
+          - target: google_apis
+            api-level: 29
+        include:
+          - target: google_apis
+            api-level: 30
+            arch: x86_64
```

Comment:
lots of changes in this section, why?

Reply:
I see you added api 30 device test before?

Reply:
yeah, I was added more emulators and tried various options but ran into flakes so left it out so that I could move forward and merge for now.

Reply:
So I need to comment it out?

Reply:
API 29 was failed as well :)